internal/armada/server/lease.go

```diff
@@ -463,102 +467,56 @@ func (q *AggregatedQueueServer) getJobs(ctx context.Context, req *api.StreamingL
 		req.ClusterId,
 		req.Pool,
+		q.schedulingConfig.Preemption.PriorityClasses,
+		q.schedulingConfig.Preemption.DefaultPriorityClass,
+		q.schedulingConfig.ResourceScarcity,
+		// May need priority factors for inactive queues for rescheduling evicted jobs.
+		priorityFactorByQueue,
+		nodeDb.TotalResources(),
+		allocatedByQueueForPool,
+	)
+	constraints := schedulerconstraints.SchedulingConstraintsFromSchedulingConfig(
 		schedulerobjects.ResourceList{Resources: req.MinimumJobSize},
 		q.schedulingConfig,
-		schedulerobjects.ResourceList{Resources: totalCapacity},
 	)
-
-	var preemptedJobs []scheduler.LegacySchedulerJob
-	var scheduledJobs []scheduler.LegacySchedulerJob
-	var schedulingContext *scheduler.SchedulingContext
-	if q.schedulingConfig.Preemption.PreemptToFairShare {
-		rescheduler := scheduler.NewRescheduler(
-			*constraints,
-			q.schedulingConfig,
-			&SchedulerJobRepositoryAdapter{
-				r: q.jobRepository,
-			},
-			nodeDb,
-			// May need priority factors for inactive queues for rescheduling evicted jobs.
-			priorityFactorByQueue,
-			allocatedByQueueForPool,
-			nodeIdByJobId,
-			jobIdsByGangId,
-			gangIdByJobId,
-		)
-		if q.schedulingConfig.EnableAssertions {
-			rescheduler.EnableAssertions()
-		}
-		result, err := rescheduler.Schedule(ctx)
-		if err != nil {
-			return nil, err
-		}
-		preemptedJobs = result.PreemptedJobs
-		scheduledJobs = result.ScheduledJobs
-		nodeIdByJobId = result.NodeIdByJobId
-		schedulingContext = result.SchedulingContext
-	} else {
-		schedulerQueues := make([]*scheduler.Queue, len(activeQueues))
-		for i, apiQueue := range activeQueues {
-			jobIterator, err := scheduler.NewQueuedJobsIterator(
-				ctx,
-				apiQueue.Name,
-				&SchedulerJobRepositoryAdapter{
-					r: q.jobRepository,
-				},
-			)
-			if err != nil {
-				return nil, err
-			}
-			queue, err := scheduler.NewQueue(
-				apiQueue.Name,
-				priorityFactorByActiveQueue[apiQueue.Name],
-				jobIterator,
-			)
-			if err != nil {
-				return nil, err
-			}
-			schedulerQueues[i] = queue
-		}
-		sched, err := scheduler.NewLegacyScheduler(
-			ctx,
-			*constraints,
-			nodeDb,
-			schedulerQueues,
-			allocatedByQueueForPool,
-		)
-		if err != nil {
-			return nil, err
-		}
-
-		// Log initial scheduler state.
-		log.Info("LegacyScheduler:\n" + sched.String())
-
-		// Run the scheduler.
-		result, err := sched.Schedule(ctx)
-		if err != nil {
-			return nil, err
-		}
-		preemptedJobs = result.PreemptedJobs
-		scheduledJobs = result.ScheduledJobs
-		nodeIdByJobId = result.NodeIdByJobId
-		schedulingContext = result.SchedulingContext
+	sch := scheduler.NewPreemptingQueueScheduler(
+		sctx,
+		constraints,
+		q.schedulingConfig.Preemption.NodeEvictionProbability,
+		q.schedulingConfig.Preemption.NodeOversubscriptionEvictionProbability,
+		&SchedulerJobRepositoryAdapter{
+			r: q.jobRepository,
+		},
+		nodeDb,
+		nodeIdByJobId,
+		jobIdsByGangId,
+		gangIdByJobId,
+	)
+	result, err := sch.Schedule(
+		ctxlogrus.ToContext(
```

Comment:
is it right to use the background context here?  I though previously we used the grpc context which meant a) we would timeout, b) we would automatically stop scheduling if the grpc channel died.

Reply:
You're right. That's an error. Fixed.